### PR TITLE
feat: add support for implicit exact Flow types

### PR DESCRIPTION
**Pull request by mrtnzlml**
This commit adds a support for implicit exact Flow types. This is especially useful when using `exact_by_default=true` Flow option and migrating towards `{ }` and `{ ... }` syntax (Flow standard is now to use `{ }` for strict objects and `{ key: value, ... }` for open objects).

Closes: https://github.com/gajus/eslint-plugin-flowtype/issues/467

TODO:

- [x] rebase after https://github.com/gajus/eslint-plugin-flowtype/pull/470
- [x] regenerate docs

### CODE REVIEW

File: tests/rules/assertions/requireReadonlyReactProps.js [hunk 1]
```diff
@@ -194,5 +210,13 @@ export default {
...
       code: 'function Foo(props: {||}) { return <p /> }',
     },
+    {
+      code: 'function Foo(props: {}) { return <p /> }',
+      options: [
+        {
+          implicitExactTypes: true,
+        },
+      ],
+    },
```
Comment (Brianzchen) [inline tests/rules/assertions/requireReadonlyReactProps.js:220]: I'm just thinking for an extra valid case, would you wanna use `{| |}` syntax and with `implicitExactTypes: true,`. Just to test that it doesn't blow up for an edge case?

File: src/rules/requireReadonlyReactProps.js [hunk 2]
```diff
@@ -21,13 +33,16 @@ const isReactComponent = (node) => {
...
 
-  // we consider `{||}` to be ReadOnly since it's exact AND has no props
-  if (node.exact && node.properties.length === 0) {
+  // we consider `{}` to be ReadOnly since it's exact AND has no props (when `implicitExactTypes=true`)
+  // we consider `{||}` to be ReadOnly since it's exact AND has no props (when `implicitExactTypes=false`)
+  if (implicitExactTypes === true && node.exact === false && node.properties.length === 0) {
+    return true;
+  } else if (node.exact === true && node.properties.length === 0) {
     return true;
   }
```
Comment (Brianzchen) [inline src/rules/requireReadonlyReactProps.js:47]: I'm wondering if this is nicer and more readable, the close duplication in syntax of your two conditions took me a while to notice that `node.exact` `===` `true/false` in both conditions
```suggestion
  if (node.properties.length === 0) {
    if (node.exact === false && implicitExactTypes === true) {
      return true;
    }
    if (node.exact === true) {
      return true;
    }
  }
```

File: tests/rules/assertions/requireReadonlyReactProps.js (hunk 1 above)
Comment (mrtnzlml) [reply tests/rules/assertions/requireReadonlyReactProps.js:244]: Sure, good idea. 👍

**Review by Brianzchen — APPROVED** [review]
LGTM

**Comment by Brianzchen** [comment]
@gajus

**Comment by Brianzchen** [comment]
@gajus bump

**Comment by mrtnzlml** [comment]
Hi, is there anything I can improve in this PR to make it mergeable?

I would also like to offer my help with the maintenance of this project if needed. We are using this plugin for years and it's always been a pleasure. I would not like to see it slowly die. 😎

**Comment by gajus** [comment]
I cannot recall the reason I did not merge this the first few times it was bumped. Sorry about that.
